weld/codegen/llvm2/cmp.rs

```diff
@@ -0,0 +1,351 @@
+//! Generates code to compare values using a comparison function.
+
+extern crate llvm_sys;
+
+use std::ffi::{CStr, CString};
+
+use ast::Type;
+use super::vector::VectorExt;
+use ast::ScalarKind::{I32, I64};
+use error::*;
+use sir::FunctionId;
+
+use super::llvm_exts::*;
+use super::llvm_exts::LLVMExtAttribute::*;
+
+use self::llvm_sys::prelude::*;
+use self::llvm_sys::core::*;
+use self::llvm_sys::LLVMLinkage;
+use self::llvm_sys::LLVMIntPredicate::*;
+
+use super::CodeGenExt;
+use super::LlvmGenerator;
+
+use ast::BinOpKind::*;
+use codegen::llvm2::numeric::gen_binop;
+use codegen::llvm2::SIR_FUNC_CALL_CONV;
+
+/// Returns whether a value can be compared with libc's `memcmp`.
+trait SupportsMemCmp {
+    fn supports_memcmp(&self) -> bool;
+}
+
+/// Only i8 or smaller values can be compared using `memcmp`.
+impl SupportsMemCmp for Type {
+    fn supports_memcmp(&self) -> bool {
+        use ast::Type::*;
+        // Structs do not support memcmp because they may be padded.
+        match *self {
+            Scalar(ref kind) => {
+                use ast::ScalarKind::*;
+                match *kind {
+                    Bool | I8 | U8 | U16 | U32 | U64 => true,
+                    _ => false,
+                }
+            },
+            _ => false,
+        }
+    }
+}
+
+/// Trait for generating comparison code.
+pub trait GenCmp {
+    /// Generates a comparator for a type.
+    ///
+    /// The comparators are over pointers of the type, e.g., a comparator for `i32`
+    /// has the type signature `i1 (i32*, i32*)`. This method returns the generated function. The
+    /// function returns -1 if the first element is smaller, 0 if the elements are equal, and 1 if
+    /// the first element is larger.
+    unsafe fn gen_cmp_fn(&mut self, ty: &Type) -> WeldResult<LLVMValueRef>; 
+
+    /// Generates an opaque comparator for a type.
+    ///
+    /// Opaque comparators are linked externally and operate over opaque (i.e., `void*`)
+    /// arguments. These are used, e.g., as an argument to qsort. This method returns the
+    /// generated function.
+    unsafe fn gen_opaque_cmp_fn(&mut self, ty: &Type) -> WeldResult<LLVMValueRef>;
```

Comment:
Delete this declaration and its implementation: I actually don't think this is necessary since you have the `gen_custom_cmp` thing, which does all the visibility stuff (it's also not used anywhere) since gen_custom_cmp will internally generate the `cmp` code and wrap it to be externally visible already.

Reply:
Sure, wasn't sure if we wanted to leave it in for potential future external calls